### PR TITLE
Support Rate Limit Rule For Application Gateway WAF Custom Rule

### PR DESCRIPTION
**Pull request by Shawnli222**
## Description

This powershell PR is supporting Application Gateway WAF Rate Limit feature. This feature is built on the existing Custom Rule framework. A new Custom Rule type "RateLimitRule" has been added. Along with this, a new field GroupByUserSession has been added to allow users to configure how to identify unique user sessions or to rate limit requests to an url endpoint by choosing "None"

Commandlet Review PR: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1322

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Network/Network/FirewallPolicy/FirewallCustomRule/AzureApplicationGatewayFirewallCustomRuleBase.cs [hunk 1]
```diff
@@ -68,6 +88,11 @@ public override void ExecuteCmdlet()
             {
                 this.State = "Enabled";
             }
+
+            if (!this.MyInvocation.BoundParameters.ContainsKey("RateLimitThreshold"))
```
Comment (tejasshah7) [inline src/Network/Network/FirewallPolicy/FirewallCustomRule/AzureApplicationGatewayFirewallCustomRuleBase.cs:92]: Why is this needed? Was this to get around the Swagger issue? If so, can this be removed now?

File: src/Network/Network/Network.csproj [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.PowerShell.AutoMapper" Version="6.2.2" />
+    <PackageReference Include="Microsoft.Azure.Management.Network" Version="26.0.0" />
```
Comment (tejasshah7) [inline src/Network/Network/Network.csproj:17]: You will probably need to update ALL the csproj file that require this update? Might want to confirm with PS team though.

File: src/Network/Network/Network.csproj [hunk 3]
```diff
@@ -41,9 +42,6 @@
     <Folder Include="RouteServerPeer\" />
   </ItemGroup>
   
-  <ItemGroup>
-    <ProjectReference Include="..\Network.Management.Sdk\Network.Management.Sdk.csproj" />
```
Comment (tejasshah7) [inline src/Network/Network/Network.csproj:45]: Why is this removed?

File: src/Network/Network/help/New-AzApplicationGatewayGroupByVariable.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Network.dll-Help.xml
+Module Name: Az.Network
+online version: https://learn.microsoft.com/powershell/module/az.network/new-azapplicationgatewaygroupbyvariable
+schema: 2.0.0
+---
+
+# New-AzApplicationGatewayGroupByVariable
+
+## SYNOPSIS
+Creates a new GroupByVariable for the application gateway GroupByUserSession.
```
Comment (tejasshah7) [inline src/Network/Network/help/New-AzApplicationGatewayGroupByVariable.md:11]: applicaiton gateway "firewall custom rule" GroupByUserSession.

File: src/Network/Network/Network.csproj (hunk 3 above)
Comment (Shawnli222) [reply src/Network/Network/Network.csproj:45]: These are instructions by Mikhail to enable network sdk.

File: src/Network/Network/Network.csproj (hunk 2 above)
Comment (Shawnli222) [reply src/Network/Network/Network.csproj:17]: These are instructions by Mikhail to enable network sdk. Only this is needed.

File: src/Network/Network/FirewallPolicy/FirewallCustomRule/AzureApplicationGatewayFirewallCustomRuleBase.cs (hunk 1 above)
Comment (Shawnli222) [reply src/Network/Network/FirewallPolicy/FirewallCustomRule/AzureApplicationGatewayFirewallCustomRuleBase.cs:92]: yes, it was used to get around swagger, will remove it

File: src/Network/Network/Network.csproj (hunk 2 above)
Comment (tejasshah7) [reply src/Network/Network/Network.csproj:17]: You can see many other files that need to be updated from this PR:
image

Probably PS team might ask you to do so.

Comment (Shawnli222) [reply src/Network/Network/Network.csproj:17]: Actally I had those files in the PR, then the PS team asked me to remove them, will check again tonight

File: src/Network/Network/help/New-AzApplicationGatewayFirewallGroupByVariable.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+external help file: Microsoft.Azure.PowerShell.Cmdlets.Network.dll-Help.xml
+Module Name: Az.Network
+online version: https://learn.microsoft.com/powershell/module/az.network/new-azapplicationgatewaygroupbyvariable
+schema: 2.0.0
+---
+
+# New-AzApplicationGatewayFirewallGroupByVariable
+
+## SYNOPSIS
+Creates a new GroupByVariable for the application gateway Firewall GroupByUserSession.
```
Comment (tejasshah7) [inline src/Network/Network/help/New-AzApplicationGatewayFirewallGroupByVariable.md:11]: Match this to previous definition and cases. Above we say application gateway firewall custom rule

File: src/Network/Network/help/New-AzApplicationGatewayFirewallGroupByVariable.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+## SYNTAX
+
+```
+New-AzApplicationGatewayFirewallGroupByVariable -VariableName<String> 
+ [-DefaultProfile <IAzureContextContainer>] [<CommonParameters>]
+```
+
+## DESCRIPTION
+The **New-AzApplicationGatewayGroupByVariable** creates a new GroupByVariable for the application gateway Firewall GroupByUserSession.
```
Comment (tejasshah7) [inline src/Network/Network/help/New-AzApplicationGatewayFirewallGroupByVariable.md]: This too, match like previous one to be consistent.

**Review by tejasshah7 — APPROVED** [review]

File: src/Network/Network/ChangeLog.md [hunk 7]
```diff
@@ -44,6 +44,11 @@
 * Blocked some regions when creating/updating Basic Sku firewall
 * Fixed bugs related to auto learn IP prefixes and Snat
 * Updated multi-auth to be supported when both OpenVPN and IkeV2 protocols are used for VNG and VWAN VPN
+* Added new cmdlets to support Rate Limiting Rule for Application Gateway WAF
+    - 'New-AzApplicationGatewayFirewallCustomRuleGroupByUserSession',
```
Comment (wyunchi-ms) [inline src/Network/Network/ChangeLog.md:48]: Please move this to section `Upcoming Release`

**Review by wyunchi-ms — APPROVED** [review]
